pkg/pillar/cmd/tpmmgr/tpmmgr.go

```diff
@@ -341,19 +371,130 @@ func readCredentials() error {
 	return nil
 }
 
-func printCapability() error {
+//till we have next version of go-tpm released, use this
+const (
+	tpmPropertyManufacturer = 0x105
+	tpmPropertyVendorStr1   = 0x106
+	tpmPropertyVendorStr2   = 0x107
+	tpmPropertyFirmVer1     = 0x10b
+	tpmPropertyFirmVer2     = 0x10c
+)
+
+func getModelName(vendorValue1 uint32, vendorValue2 uint32) string {
+	uintToByteArr := func(value uint32) []byte {
+		get8 := func(val uint32, offset uint32) uint8 {
+			return (uint8)((val >> ((3 - offset) * 8)) & 0xff)
+		}
+		var i uint32
+		var bytes []byte
+		for i = 0; i < uint32(unsafe.Sizeof(value)); i++ {
+			c := get8(value, i)
+			bytes = append(bytes, c)
+		}
+		return bytes
+	}
+	var model []byte
+	model = append(model, uintToByteArr(vendorValue1)...)
+	model = append(model, uintToByteArr(vendorValue2)...)
+	return string(model)
+}
+
+func getFirmwareVersion(v1 uint32, v2 uint32) string {
+	get16 := func(val uint32, offset uint32) uint16 {
+		return uint16((val >> ((1 - offset) * 16)) & 0xFFFF)
+	}
+	return fmt.Sprintf("%d.%d.%d.%d", get16(v1, 0), get16(v1, 1),
+		get16(v2, 0), get16(v2, 1))
+}
+
+func getTpmProperty(propID uint32) (uint32, error) {
 	lockTpmAccess()
 	defer unlockTpmAccess()
 
 	rw, err := tpm2.OpenTPM(TpmDevicePath)
 	if err != nil {
-		return err
+		return 0, err
 	}
 	defer rw.Close()
-	l, _, err := tpm2.GetCapability(rw, tpm2.CapabilityTPMProperties,
-		1, uint32(0x105))
-	fmt.Println(l)
-	return err
+
+	v, _, err := tpm2.GetCapability(rw, tpm2.CapabilityTPMProperties, 1, propID)
+	if err != nil {
+		return 0, err
+	}
+	prop, ok := v[0].(tpm2.TaggedProperty)
+	if !ok {
+		return 0, fmt.Errorf("Unable to fetch property %d", propID)
+	}
+	return prop.Value, nil
+}
+
+//FetchTpmSwStatus returns states reflecting SW usage of TPM
+func FetchTpmSwStatus() info.HwSecurityModuleStatus {
+	_, err := os.Stat(TpmDevicePath)
+	if err != nil {
+		//No TPM found on this system
+		return info.HwSecurityModuleStatus_NOTFOUND
+	}
+
+	if IsTpmEnabled() {
+		//TPM is found and is used by software
+		return info.HwSecurityModuleStatus_ENABLED
+	}
+
+	//TPM is found but not being used by software
+	return info.HwSecurityModuleStatus_DISABLED
+}
+
+//FetchTpmHwInfo returns TPM Hardware properties in a string
+func FetchTpmHwInfo() (string, error) {
+
+	//If we had done this earlier, return the last result
+	if tpmHwInfo != "" {
+		return tpmHwInfo, nil
+	}
+
+	//Take care of non-TPM platforms
+	_, err := os.Stat(TpmDevicePath)
+	if err != nil {
+		tpmHwInfo = "Not Available"
+		return tpmHwInfo, nil
+	}
+
+	//First time. Fetch it from TPM and cache it.
```

Comment:
Where are you caching it? Looks like you'll redo this each time the device sends a device info message, which might be excessive.

If it is useful to cache perhaps handlemetrics can do the caching? Or tpmmgr?